lookout/style/format/analyzer.py

```diff
@@ -289,6 +238,100 @@ def train(cls, ptr: ReferencePointer, config: Mapping[str, Any], data_request_st
 
+    def _generate_file_comments(self, lang, file, fe, feature_extractor_output, rules):
+        file_comments = []
+        X, y, (vnodes_y, vnodes, vnode_parents, node_parents) = feature_extractor_output
+        y_pred, rule_winners = rules.predict(X=X, vnodes_y=vnodes_y, vnodes=vnodes,
+                                             feature_extractor=fe)
+        if self.config["uast_break_check"]:
+            y, y_pred, vnodes_y, safe_preds = filter_uast_breaking_preds(
+                y=y, y_pred=y_pred, vnodes_y=vnodes_y, vnodes=vnodes, files={file.path: file},
+                feature_extractor=fe, client=self.client, vnode_parents=vnode_parents,
+                node_parents=node_parents, log=self._log)
+        assert len(y) == len(y_pred)
+        code_generator = CodeGenerator(fe, skip_errors=True)
+        new_vnodes = code_generator.apply_predicted_y(vnodes, vnodes_y, y_pred)
+        code_lines = file.content.decode("utf-8", "replace").splitlines()
+        for line_number, line in self._group_line_nodes(
+                y, y_pred, vnodes_y, new_vnodes, rule_winners):
+            line_ys, line_ys_pred, line_vnodes_y, new_line_vnodes, line_winners = line
+            new_code_line = code_generator.generate(
+                new_line_vnodes, "local").lstrip("\n").splitlines()[0]
+            descriptions = []
+            for line_yi, line_y_predi, line_vnode_y, line_winner in zip(
+                    line_ys, line_ys_pred, line_vnodes_y, line_winners):
+                if line_yi == line_y_predi:
+                    continue
+                descriptions.append(
+                    (line_winner, get_error_description(line_vnode_y, line_y_predi, fe)))
+            text = self.report_template.render(
+                config=self.config,
+                lang=lang,
+                line_number=line_number,
+                code_lines=code_lines,
+                new_code_line=new_code_line,
+                descriptions=descriptions,
+                describe_rule=describe_rule,
+                feature_extractor=fe,
+                rules=rules.rules,
+            )
+
+            file_comments.append(
+                generate_comment(
+                    filename=file.path,
+                    line=line_number,
+                    text=text,
+                    confidence=self._get_comment_confidence(
+                        line_ys, line_ys_pred, line_winners, rules)))
+
+        return file_comments
+
+    @staticmethod
+    def _get_comment_confidence(line_ys, line_ys_pred, line_winners, rules):
+        confidence = 0
+        confidence_count = 0
+        for yi, y_predi, winner in zip(line_ys, line_ys_pred, line_winners):
+            if yi != y_predi:
+                confidence += rules.rules[winner].stats.conf
+                confidence_count += 1
+        return int(round(confidence * 100 / confidence_count))
+
+    @staticmethod
+    def _group_line_nodes(y, y_pred, vnodes_y, new_vnodes, rule_winners):
```

Comment:
I see that we are doing many complex actions here... Can you please add a docstring which describes the hell inside?